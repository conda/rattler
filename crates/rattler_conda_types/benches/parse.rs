use criterion::{criterion_group, criterion_main, Criterion};
use rattler_conda_types::Version;
use std::hint::black_box;

fn criterion_benchmark(c: &mut Criterion) {
    c.bench_function("parse simple version", |b| {
        b.iter(|| black_box("3.11.4").parse::<Version>());
    });
    c.bench_function("parse complex version", |b| {
        b.iter(|| black_box("1!1.0b2.post345.dev456+3.2.20.rc3").parse::<Version>());
    });
    c.bench_function("parse logical constraint", |b| {
        b.iter(|| black_box(">=3.1").parse::<Version>());
    });
    c.bench_function("parse wildcard constraint", |b| {
        b.iter(|| black_box("3.1.*").parse::<Version>());
    });
    c.bench_function("parse simple version spec", |b| {
        b.iter(|| black_box(">=3.1").parse::<Version>());
    });
    c.bench_function("parse complex version spec", |b| {
        b.iter(|| black_box("(>=2.1.0,<3.0)|(~=3.2.1,~3.2.2.1)|(==4.1)").parse::<Version>());
    });
}

criterion_group!(benches, criterion_benchmark);
criterion_main!(benches);
