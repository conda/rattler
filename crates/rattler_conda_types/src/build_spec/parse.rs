//! this module supports the parsing features of the build number spec

use super::{BuildNumber, BuildNumberSpec, OrdOperator};

use nom::{bytes::complete::take_while1, character::complete::digit1, Finish, IResult};
use std::str::FromStr;
use thiserror::Error;

impl FromStr for BuildNumberSpec {
    type Err = ParseBuildNumberSpecError;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        match Self::parser(s).finish()? {
            ("", spec) => Ok(spec),
            (_, _) => Err(ParseBuildNumberSpecError::ExpectedEof),
        }
    }
}

impl BuildNumberSpec {
    /// Parses a build number spec, string representation is optional operator preceding whole number
    pub fn parser(input: &str) -> IResult<&str, BuildNumberSpec, ParseBuildNumberSpecError> {
        // Parse the optional preceding operator
        let (input, op) = match OrdOperator::parser(input) {
            Err(
                nom::Err::Failure(ParseOrdOperatorError::InvalidOperator(op))
                | nom::Err::Error(ParseOrdOperatorError::InvalidOperator(op)),
            ) => {
                return Err(nom::Err::Failure(
                    ParseBuildNumberSpecError::InvalidOperator(
                        ParseOrdOperatorError::InvalidOperator(op),
                    ),
                ))
            }
            Err(nom::Err::Error(_)) => (input, None),
            Ok((rest, op)) => (rest, Some(op)),
            _ => unreachable!(),
        };

        let (rest, build_num) = digit1(input)
            .map(|(rest, digits): (&str, &str)| {
                (
                    rest,
                    digits
                        .parse::<BuildNumber>()
                        .expect("nom found at least 1 digit(s)"),
                )
            })
            .map_err(|_err: nom::Err<nom::error::Error<&str>>| {
                nom::Err::Error(ParseBuildNumberSpecError::InvalidBuildNumber(
                    ParseBuildNumberError,
                ))
            })?;

        match op {
            Some(op) => Ok((rest, BuildNumberSpec::new(op, build_num))),
            None => Ok((rest, BuildNumberSpec::new(OrdOperator::Eq, build_num))),
        }
    }
}

/// Possible errors when parsing the [`OrdOperator`] which precedes the digits in a build number
/// spec.
#[derive(Debug, Clone, Error, Eq, PartialEq)]
pub enum ParseOrdOperatorError {
    /// Indicates that operator symbols were captured,
    /// but not interpretable as an `OrdOperator`
    #[error("invalid operator '{0}'")]
    InvalidOperator(String),
    /// Indicates no symbol sequence found for `OrdOperator`s,
    /// callers should expect explicit operators
    #[error("expected version operator")]
    ExpectedOperator,
    /// Indicates that there was data after an operator was read,
    /// callers should handle this if they expect input to end with the operator
    #[error("expected EOF")]
    ExpectedEof,
}

/// Simplified error when parsing the digits into a build number: u64 in a build number spec
#[derive(Debug, Clone, Eq, PartialEq, Error)]
#[error("could not parse build number")]
pub struct ParseBuildNumberError;

/// Composition of possible errors when parsing the spec for build numbers
#[allow(clippy::enum_variant_names, missing_docs)]
#[derive(Debug, Clone, Error, Eq, PartialEq)]
pub enum ParseBuildNumberSpecError {
    #[error("invalid version: {0}")]
    InvalidBuildNumber(#[source] ParseBuildNumberError),
    #[error("invalid version constraint: {0}")]
    InvalidOperator(#[source] ParseOrdOperatorError),
    #[error("expected EOF")]
    ExpectedEof,
}

impl FromStr for OrdOperator {
    type Err = ParseOrdOperatorError;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        match Self::parser(s).finish()? {
            ("", spec) => Ok(spec),
            (_, _) => Err(ParseOrdOperatorError::ExpectedEof),
        }
    }
}

impl OrdOperator {
    /// Parses an operator representing [`PartialOrd`] compares, returns an error if the operator is
    /// not recognized or not found.
    fn parser(input: &str) -> IResult<&str, OrdOperator, ParseOrdOperatorError> {
        // Take anything that looks like an operator.
        let (rest, operator_str) = take_while1(|c| "=!<>".contains(c))(input).map_err(
            |_err: nom::Err<nom::error::Error<&str>>| {
                nom::Err::Error(ParseOrdOperatorError::ExpectedOperator)
            },
        )?;

        let op = match operator_str {
            "==" => OrdOperator::Eq,
            "!=" => OrdOperator::Ne,
            "<=" => OrdOperator::Le,
            ">=" => OrdOperator::Ge,
            "<" => OrdOperator::Lt,
            ">" => OrdOperator::Gt,
            _ => {
                return Err(nom::Err::Failure(ParseOrdOperatorError::InvalidOperator(
                    operator_str.to_string(),
                )))
            }
        };

        Ok((rest, op))
    }
}

#[cfg(test)]
mod test {
    use super::{BuildNumberSpec, OrdOperator, ParseOrdOperatorError};

    use nom::Finish;

    #[test]
    fn parse_operator_from_spec() {
        let test_params = vec![
            (">3.1", OrdOperator::Gt),
            (">=3.1", OrdOperator::Ge),
            ("<3.1", OrdOperator::Lt),
            ("<=3.1", OrdOperator::Le),
            ("==3.1", OrdOperator::Eq),
            ("!=3.1", OrdOperator::Ne),
        ];

        for (s, op) in test_params {
            assert_eq!(OrdOperator::parser(s), Ok(("3.1", op)));
        }

        assert_eq!(
            OrdOperator::parser("<==>3.1"),
            Err(nom::Err::Failure(ParseOrdOperatorError::InvalidOperator(
                "<==>".to_string()
            )))
        );
        assert_eq!(
            OrdOperator::parser("3.1"),
            Err(nom::Err::Error(ParseOrdOperatorError::ExpectedOperator))
        );
    }

    #[test]
    fn parse_spec() {
        let test_params = vec![
            (">1", OrdOperator::Gt),
            (">=1", OrdOperator::Ge),
            ("<1", OrdOperator::Lt),
            ("<=1", OrdOperator::Le),
            ("==1", OrdOperator::Eq),
            ("!=1", OrdOperator::Ne),
        ];

        for (s, op) in test_params {
            assert_eq!(
                BuildNumberSpec::parser(s),
                Ok(("", BuildNumberSpec::new(op, 1)))
            );
        }

        assert_eq!(
            BuildNumberSpec::parser(">=1.1"),
            Ok((".1", BuildNumberSpec::new(OrdOperator::Ge, 1)))
        );

        assert!(BuildNumberSpec::parser(">=build3").finish().is_err());
    }
}
