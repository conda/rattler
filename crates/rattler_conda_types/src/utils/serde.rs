use chrono::{DateTime, Utc};
use fxhash::FxHashMap;
use serde::de::Error as _;
use serde::ser::Error;
use serde::{Deserialize, Deserializer, Serialize, Serializer};
use serde_with::{DeserializeAs, SerializeAs};
use std::collections::BTreeMap;
use std::marker::PhantomData;
use std::path::{Path, PathBuf};
use url::Url;

/// A helper struct that serializes Paths in a normalized way.
/// - Backslashes are replaced with forward-slashes.
pub(crate) struct NormalizedPath;

impl<P: AsRef<Path>> SerializeAs<P> for NormalizedPath {
    fn serialize_as<S>(source: &P, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: Serializer,
    {
        match source.as_ref().to_str() {
            Some(s) => s.replace('\\', "/").serialize(serializer),
            None => Err(S::Error::custom("path contains invalid UTF-8 characters")),
        }
    }
}

impl<'de> DeserializeAs<'de, PathBuf> for NormalizedPath {
    fn deserialize_as<D>(deserializer: D) -> Result<PathBuf, D::Error>
    where
        D: Deserializer<'de>,
    {
        PathBuf::deserialize(deserializer)
    }
}

/// Deserialize a sequence into `Vec<T>` but filter `None` values.
pub(crate) struct VecSkipNone<T>(PhantomData<T>);

impl<'de, T, I> DeserializeAs<'de, Vec<T>> for VecSkipNone<I>
where
    I: DeserializeAs<'de, Vec<Option<T>>>,
{
    fn deserialize_as<D>(deserializer: D) -> Result<Vec<T>, D::Error>
    where
        D: Deserializer<'de>,
    {
        Ok(I::deserialize_as(deserializer)?
            .into_iter()
            .flatten()
            .collect())
    }
}

/// A helper type parser that tries to parse Urls that could be malformed.
pub(crate) struct LossyUrl;

impl<'de> DeserializeAs<'de, Option<Url>> for LossyUrl {
    fn deserialize_as<D>(deserializer: D) -> Result<Option<Url>, D::Error>
    where
        D: Deserializer<'de>,
    {
        let str = match Option::<String>::deserialize(deserializer)? {
            Some(url) => url,
            None => return Ok(None),
        };
        let url = match Url::parse(&str) {
            Ok(url) => url,
            Err(e) => {
                tracing::warn!("unable to parse '{}' as an URL: {e}. Skipping...", str);
                return Ok(None);
            }
        };
        Ok(Some(url))
    }
}

/// A helper type that parses a string either as a string or a vector of strings.
pub(crate) struct MultiLineString;

impl<'de> DeserializeAs<'de, String> for MultiLineString {
    fn deserialize_as<D>(deserializer: D) -> Result<String, D::Error>
    where
        D: Deserializer<'de>,
    {
        #[derive(Deserialize)]
        #[serde(untagged)]
        enum Inner {
            String(String),
            Multi(Vec<String>),
        }

        Ok(match Inner::deserialize(deserializer)? {
            Inner::String(s) => s,
            Inner::Multi(s) => s.join("\n"),
        })
    }
}

pub(crate) struct Timestamp;

impl<'de> DeserializeAs<'de, chrono::DateTime<chrono::Utc>> for Timestamp {
    fn deserialize_as<D>(deserializer: D) -> Result<chrono::DateTime<chrono::Utc>, D::Error>
    where
        D: Deserializer<'de>,
    {
        let timestamp = i64::deserialize(deserializer)?;

        // Convert from milliseconds to seconds
        let microseconds = if timestamp > 253_402_300_799 {
            timestamp * 1_000
        } else {
            timestamp * 1_000_000
        };

        // Convert the timestamp to a UTC timestamp
        chrono::DateTime::from_timestamp_micros(microseconds)
            .ok_or_else(|| D::Error::custom("got invalid timestamp, timestamp out of range"))
    }
}

impl SerializeAs<chrono::DateTime<chrono::Utc>> for Timestamp {
    fn serialize_as<S>(source: &DateTime<Utc>, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: Serializer,
    {
        // Convert the date to a timestamp
        let timestamp: i64 = source.timestamp_millis();

        // Determine the precision of the timestamp.
        let timestamp = if timestamp % 1000 == 0 {
            timestamp / 1000
        } else {
            timestamp
        };

        // Serialize the timestamp
        timestamp.serialize(serializer)
    }
}

/// A helper struct to deserialize types from a string without checking the string.
pub struct DeserializeFromStrUnchecked;

/// A helper function used to sort map alphabetically when serializing.
pub(crate) fn sort_map_alphabetically<T: Serialize, S: serde::Serializer>(
    value: &FxHashMap<String, T>,
    serializer: S,
) -> Result<S::Ok, S::Error> {
    value
        .iter()
        .collect::<BTreeMap<_, _>>()
        .serialize(serializer)
}
