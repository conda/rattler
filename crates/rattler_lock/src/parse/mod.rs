mod deserialize;
mod models;
mod serialize;
mod v3;

use std::str::FromStr;

use rattler_conda_types::Platform;
use serde::de::Error;
use serde_yaml::Value;
use v3::parse_v3_or_lower;

use super::{LockFile, UrlOrPath};
use crate::{file_format_version::FileFormatVersion, parse::deserialize::parse_from_document_v5};

#[allow(missing_docs)]
#[derive(Debug, thiserror::Error)]
pub enum ParseCondaLockError {
    #[error(transparent)]
    IoError(#[from] std::io::Error),

    #[error(transparent)]
    ParseError(#[from] serde_yaml::Error),

    #[error("found newer lockfile format version {lock_file_version}, but only up to including version {max_supported_version} is supported")]
    IncompatibleVersion {
        lock_file_version: u64,
        max_supported_version: FileFormatVersion,
    },

    #[error("environment {0} and platform {1} refers to a package that does not exist: {2}")]
    MissingPackage(String, Platform, UrlOrPath),

    #[error(transparent)]
    InvalidPypiPackageName(#[from] pep508_rs::InvalidNameError),

    #[error("missing field `{0}` for package {1}")]
    MissingField(String, UrlOrPath),

    /// The location of the conda package cannot be converted to a URL
    #[error(transparent)]
    LocationToUrlConversionError(#[from] file_url::FileURLParseError),
}

impl FromStr for LockFile {
    type Err = ParseCondaLockError;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        // First parse the document to a `serde_yaml::Value`.
        let document: Value = serde_yaml::from_str(s).map_err(ParseCondaLockError::ParseError)?;

        // Read the version number from the document
        let version: FileFormatVersion = document
            .get("version")
            .ok_or_else(|| {
                ParseCondaLockError::ParseError(serde_yaml::Error::custom(
                    "missing `version` field in lock file",
                ))
            })
            .and_then(|v| {
                let v = v.as_u64().ok_or_else(|| {
                    ParseCondaLockError::ParseError(serde_yaml::Error::custom(
                        "`version` field in lock file is not an integer",
                    ))
                })?;

                FileFormatVersion::try_from(v)
            })?;

        if version <= FileFormatVersion::V3 {
            parse_v3_or_lower(document, version)
        } else if version <= FileFormatVersion::V5 {
            parse_from_document_v5(document, version)
        } else {
            deserialize::parse_from_document_v6(document, version)
        }
    }
}

/// A helper struct to differentiate between the serde code paths for different
/// versions.
struct V5;

/// A helper struct to differentiate between the serde code paths for different
/// versions.
struct V6;

#[cfg(test)]
mod test {
    use std::path::Path;

    use super::*;

    #[test]
    fn test_forward_compatibility() {
        let err = LockFile::from_path(
            &Path::new(env!("CARGO_MANIFEST_DIR"))
                .join("../../test-data/conda-lock/forward-compatible-lock.yml"),
        )
        .err()
        .unwrap();

        insta::assert_snapshot!(format!("{}", err), @"found newer lockfile format version 1000, but only up to including version 6 is supported");
    }

    // This test verifies the deterministic ordering of lock files. It does so by
    // comparing the serialized YAML output of two lock files: one with the
    // original ordering and another with a shuffled ordering. The test ensures
    // that, despite the initial difference in order, the serialization process
    // results in identical YAML strings.
    #[test]
    fn test_deterministic_lock_file_ordering() {
        let lock_file_original = LockFile::from_path(
            &Path::new(env!("CARGO_MANIFEST_DIR"))
                .join("../../test-data/conda-lock/v5/stability-original.yml"),
        )
        .unwrap();
        let lock_file_shuffled = LockFile::from_path(
            &Path::new(env!("CARGO_MANIFEST_DIR"))
                .join("../../test-data/conda-lock/v5/stability-shuffled.yml"),
        )
        .unwrap();

        let output_original =
            serde_yaml::to_string(&lock_file_original).expect("could not deserialize");
        let output_shuffled =
            serde_yaml::to_string(&lock_file_shuffled).expect("could not deserialize");

        assert_eq!(output_original, output_shuffled);
    }
}
