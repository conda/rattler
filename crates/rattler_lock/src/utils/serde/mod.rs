mod match_spec_map_or_vec;
mod ordered;
mod pep440_map_or_vec;
mod timestamp;
pub(crate) mod url_or_path;
pub(crate) use match_spec_map_or_vec::MatchSpecMapOrVec;
pub(crate) use ordered::Ordered;
pub(crate) use pep440_map_or_vec::Pep440MapOrVec;
pub(crate) use timestamp::Timestamp;
