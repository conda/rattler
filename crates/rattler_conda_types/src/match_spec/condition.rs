use std::fmt::Display;

use nom::{
    branch::alt,
    bytes::complete::{tag, take_while1},
    character::complete::{char, multispace0, multispace1},
    sequence::{delimited, preceded},
    IResult, Parser,
};
use serde::Serialize;

use crate::match_spec::parse::matchspec_parser;

#[derive(Debug, Clone, PartialEq, Serialize)]
pub struct Statement {
    pub prefix: String,
    pub condition: Option<MatchSpecCondition>,
}

/// Represents a condition in a match spec, which can be a match spec itself or a logical combination
#[derive(Debug, Clone, PartialEq, Serialize, Eq, Hash)]
pub enum MatchSpecCondition {
    /// A condition on a certain match spec (e.g. `python >=3.12`)
    MatchSpec(Box<crate::MatchSpec>),
    /// A logical AND condition combining two conditions
    And(Box<MatchSpecCondition>, Box<MatchSpecCondition>),
    /// A logical OR condition combining two conditions
    Or(Box<MatchSpecCondition>, Box<MatchSpecCondition>),
}

impl Display for MatchSpecCondition {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            MatchSpecCondition::MatchSpec(ms) => write!(f, "{}", ms),
            MatchSpecCondition::And(lhs, rhs) => write!(f, "({} and {})", lhs, rhs),
            MatchSpecCondition::Or(lhs, rhs) => write!(f, "({} or {})", lhs, rhs),
        }
    }
}

// Parse identifier (alphanumeric + underscore)
fn identifier(input: &str) -> IResult<&str, &str> {
    take_while1(|c: char| c.is_alphanumeric() || c == '_')(input)
}

// Parse whitespace
fn ws(input: &str) -> IResult<&str, &str> {
    multispace0(input)
}

// Parse a matchspec by consuming until we hit a delimiter
fn matchspec_token(input: &str) -> IResult<&str, &str> {
    // Try to find the next delimiter
    let delimiters = ["and", "or", ")", "("];

    // Find the earliest delimiter
    let mut end_pos = input.len();
    for delimiter in &delimiters {
        if let Some(pos) = input.find(delimiter) {
            // Make sure it's a word boundary for "and"/"or"
            if *delimiter == "and" || *delimiter == "or" {
                // Check if it's preceded and followed by whitespace or start/end of string
                let is_word_boundary = {
                    let before_ok = pos == 0 || input.chars().nth(pos - 1).unwrap().is_whitespace();
                    let after_ok = pos + delimiter.len() >= input.len()
                        || input
                            .chars()
                            .nth(pos + delimiter.len())
                            .unwrap()
                            .is_whitespace();
                    before_ok && after_ok
                };
                if is_word_boundary {
                    end_pos = end_pos.min(pos);
                }
            } else {
                end_pos = end_pos.min(pos);
            }
        }
    }

    if end_pos == 0 {
        return Err(nom::Err::Error(nom::error::Error::new(
            input,
            nom::error::ErrorKind::TakeUntil,
        )));
    }

    let (matchspec_str, remaining) = input.split_at(end_pos);
    let matchspec_str = matchspec_str.trim();

    if matchspec_str.is_empty() {
        return Err(nom::Err::Error(nom::error::Error::new(
            input,
            nom::error::ErrorKind::TakeUntil,
        )));
    }

    Ok((remaining, matchspec_str))
}

// Parse a matchspec
fn matchspec(input: &str) -> IResult<&str, MatchSpecCondition> {
    let (remaining, matchspec_str) = matchspec_token(input)?;

    match matchspec_parser(matchspec_str, crate::ParseStrictness::Strict) {
        Ok(parsed_matchspec) => Ok((
            remaining,
            MatchSpecCondition::MatchSpec(Box::new(parsed_matchspec)),
        )),
        Err(_) => Err(nom::Err::Error(nom::error::Error::new(
            input,
            nom::error::ErrorKind::MapRes,
        ))),
    }
}

// Parse parenthesized condition
fn parenthesized_condition(input: &str) -> IResult<&str, MatchSpecCondition> {
    delimited((char('('), ws), parse_condition, (ws, char(')'))).parse(input)
}

// Parse primary condition (matchspec or parenthesized)
fn primary_condition(input: &str) -> IResult<&str, MatchSpecCondition> {
    alt((parenthesized_condition, matchspec)).parse(input)
}

// Parse AND expressions (higher precedence)
fn and_condition(input: &str) -> IResult<&str, MatchSpecCondition> {
    let (input, first) = primary_condition(input)?;
    let (input, rest) =
        nom::multi::many0(preceded((ws, tag("and"), multispace1), primary_condition))
            .parse(input)?;

    Ok((
        input,
        rest.into_iter().fold(first, |acc, next| {
            MatchSpecCondition::And(Box::new(acc), Box::new(next))
        }),
    ))
}

// Parse OR expressions (lower precedence)
fn or_condition(input: &str) -> IResult<&str, MatchSpecCondition> {
    let (input, first) = and_condition(input)?;
    let (input, rest) =
        nom::multi::many0(preceded((ws, tag("or"), multispace1), and_condition)).parse(input)?;

    Ok((
        input,
        rest.into_iter().fold(first, |acc, next| {
            MatchSpecCondition::Or(Box::new(acc), Box::new(next))
        }),
    ))
}

// Parse the main condition
pub(crate) fn parse_condition(input: &str) -> IResult<&str, MatchSpecCondition> {
    or_condition(input)
}

#[cfg(test)]
mod tests {
    use super::*;
    use insta::assert_yaml_snapshot;
    use nom::combinator::opt;

    fn parse_and_extract(input: &str) -> Statement {
        let result = parse_statement(input).unwrap();
        assert_eq!(result.0.trim(), ""); // Ensure no remaining input
        result.1
    }

    // Parse the entire statement
    pub fn parse_statement(input: &str) -> IResult<&str, Statement> {
        let (input, prefix) = identifier(input)?;
        let (input, _) = char(';')(input)?;
        let (input, _) = ws(input)?;
        let (input, condition) =
            opt(preceded((tag("if"), multispace1), parse_condition)).parse(input)?;

        Ok((
            input,
            Statement {
                prefix: prefix.to_string(),
                condition,
            },
        ))
    }

    #[test]
    fn test_condition_parsing_snapshots() {
        let test_cases = vec![
            "bla; if foobar or bizbaz",
            "bla; if python >=3.12 or foobar [version='3.12.*', url='https://foobar.com/bla.tar.bz2']",
            "xyz; if foobar and (bizbaz or blabla)",
            "test;",
            "simple; if single_condition",
            "complex; if a and b or c",
            "nested; if (a or b) and (c or d)",
            "deep; if a and (b or (c and d))",
            "whitespace;   if   foo   or   bar  ",
            "underscores; if foo_bar and baz_qux",
            "mixed; if (alpha and beta) or (gamma and (delta or epsilon))",
        ];

        let results: Vec<(&str, Statement)> = test_cases
            .into_iter()
            .map(|input| (input, parse_and_extract(input)))
            .collect();

        assert_yaml_snapshot!(results);
    }

    #[test]
    fn test_individual_cases() {
        // Simple OR condition
        let result = parse_and_extract("bla; if foobar or bizbaz");
        assert_yaml_snapshot!("simple_or", result);

        // Complex AND with parentheses
        let result = parse_and_extract("xyz; if foobar and (bizbaz or blabla)");
        assert_yaml_snapshot!("complex_and_with_parens", result);

        // No condition
        let result = parse_and_extract("test;");
        assert_yaml_snapshot!("no_condition", result);

        // Precedence test
        let result = parse_and_extract("prec; if a and b or c and d");
        assert_yaml_snapshot!("precedence_test", result);
    }

    #[test]
    fn test_error_cases() {
        // These should fail to parse completely
        let error_cases = vec![
            "no_semicolon if foo",
            "; if missing_prefix",
            "bad; if (unclosed_paren",
            "bad; if closed_paren)",
            "bad; if and missing_operand",
            "bad; if or missing_operand",
        ];

        for case in error_cases {
            let result = parse_statement(case);
            // These should either fail or not consume all input
            if let Ok((remaining, _)) = result {
                assert!(
                    !remaining.is_empty(),
                    "Case '{}' should have failed or left remaining input",
                    case
                );
            }
        }
    }
}
