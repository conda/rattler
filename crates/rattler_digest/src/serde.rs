//! Provides custom serialization/deserialization functions for [`Output`] of a [`Digest`]
//! Use the struct [`SerializableHash`] to easily serialize the digest.
//!
//! # Example:
//!
//! ```
//!
//! use rattler_digest::serde::SerializableHash;
//!
//! let hash = SerializableHash::<sha2::Sha256>(
//! rattler_digest::parse_digest_from_hex::<sha2::Sha256>("fe51de6107f9edc7aa4f786a70f4a883943bc9d39b3bb7307c04c41410990726").unwrap());
//! let str = serde_json::to_string(&hash).unwrap();
//! let hash: SerializableHash<sha2::Sha256> = serde_json::from_str(&str).unwrap();
//!
//! ```
//!
use digest::{Digest, Output};
use serde::de::Error;
use serde::{Deserialize, Deserializer, Serialize, Serializer};
use serde_with::{DeserializeAs, SerializeAs};
use std::borrow::Cow;
use std::fmt::LowerHex;
use std::ops::Deref;

/// Deserialize the [`Output`] of a [`Digest`].
///
/// If the deserializer is human-readable, it will parse the digest from a hex
/// string. Otherwise, it will deserialize raw bytes.
pub fn deserialize<'de, D, Dig: Digest>(deserializer: D) -> Result<Output<Dig>, D::Error>
where
    D: Deserializer<'de>,
{
    if deserializer.is_human_readable() {
        let str = Cow::<'de, str>::deserialize(deserializer)?;
        super::parse_digest_from_hex::<Dig>(str.as_ref())
            .ok_or_else(|| Error::custom("failed to parse digest"))
    } else {
        Output::<Dig>::deserialize(deserializer)
    }
}

/// Serializes the [`Output`] of a [`Digest`].
///
/// If the serializer is human-readable, it will write the digest as a hex
/// string. Otherwise, it will deserialize raw bytes.
pub fn serialize<'a, S: Serializer, Dig: Digest>(
    digest: &'a Output<Dig>,
    s: S,
) -> Result<S::Ok, S::Error>
where
    &'a Output<Dig>: LowerHex,
{
    if s.is_human_readable() {
        format!("{digest:x}").serialize(s)
    } else {
        digest.serialize(s)
    }
}

/// Wrapper type for easily serializing a Hash
#[derive(Debug, Clone)]
pub struct SerializableHash<T: Digest>(pub Output<T>);

impl<T: Digest> Serialize for SerializableHash<T>
where
    Output<T>: LowerHex,
{
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: Serializer,
    {
        serialize::<S, T>(&self.0, serializer)
    }
}

impl<'de, T: Digest + Default> Deserialize<'de> for SerializableHash<T> {
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: Deserializer<'de>,
    {
        let hash_output: Output<T> = deserialize::<D, T>(deserializer)?;
        Ok(SerializableHash(hash_output))
    }
}

impl<T: Digest> From<Output<T>> for SerializableHash<T> {
    fn from(output: Output<T>) -> Self {
        SerializableHash(output)
    }
}

impl<T: Digest> From<SerializableHash<T>> for Output<T> {
    fn from(s: SerializableHash<T>) -> Self {
        s.0
    }
}

impl<T: Digest> Deref for SerializableHash<T> {
    type Target = Output<T>;

    fn deref(&self) -> &Self::Target {
        &self.0
    }
}

impl<T: Digest> SerializeAs<Output<T>> for SerializableHash<T>
where
    for<'a> &'a Output<T>: LowerHex,
{
    fn serialize_as<S>(source: &Output<T>, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: Serializer,
    {
        serialize::<S, T>(source, serializer)
    }
}

impl<'de, T: Digest + Default> DeserializeAs<'de, Output<T>> for SerializableHash<T> {
    fn deserialize_as<D>(deserializer: D) -> Result<Output<T>, D::Error>
    where
        D: Deserializer<'de>,
    {
        deserialize::<D, T>(deserializer)
    }
}

#[cfg(test)]
mod test {
    use crate::serde::SerializableHash;

    #[test]
    pub fn test_serializable_hash() {
        let hash = SerializableHash::<sha2::Sha256>(
            crate::parse_digest_from_hex::<sha2::Sha256>(
                "fe51de6107f9edc7aa4f786a70f4a883943bc9d39b3bb7307c04c41410990726",
            )
            .unwrap(),
        );
        let str = serde_json::to_string(&hash).unwrap();
        let _hash: SerializableHash<sha2::Sha256> = serde_json::from_str(&str).unwrap();
    }
}
