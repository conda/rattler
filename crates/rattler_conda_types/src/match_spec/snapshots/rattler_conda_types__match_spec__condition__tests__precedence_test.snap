---
source: crates/rattler_conda_types/src/match_spec/condition.rs
expression: result
---
prefix: prec
condition:
  Or:
    - And:
        - MatchSpec:
            name: a
        - MatchSpec:
            name: b
    - And:
        - MatchSpec:
            name: c
        - MatchSpec:
            name: d
